File: backend/src/models/Ticket.ts
import {
  Table,
  Column,
  CreatedAt,
  UpdatedAt,
  Model,
  PrimaryKey,
  ForeignKey,
  BelongsTo,
  HasMany,
  AutoIncrement,
  Default,
  DataType,
  AllowNull,
  BelongsToMany
} from "sequelize-typescript";

import Contact from "./Contact";
import Message from "./Message";
import Queue from "./Queue";
import User from "./User";
import Whatsapp from "./Whatsapp";
import Tenant from "./Tenant";
import Step from "./Step";
import Tag from "./Tag";
import EntityTag from "./EntityTag";

@Table
class Ticket extends Model<Ticket> {
  @PrimaryKey
  @AutoIncrement
  @Column
  id: number;

  @Column({
    type: DataType.UUID,
    defaultValue: DataType.UUIDV4
  })
  uuid: string;

  @Column({ defaultValue: "pending" })
  status: string;

  @Column
  unreadMessages: number;

  @Column
  lastMessage: string;

  @Default(false)
  @Column
  isGroup: boolean;

  @CreatedAt
  createdAt: Date;

  @UpdatedAt
  updatedAt: Date;

  @ForeignKey(() => User)
  @Column
  userId: number;

  @BelongsTo(() => User)
  user: User;

  @ForeignKey(() => Contact)
  @Column
  contactId: number;

  @BelongsTo(() => Contact)
  contact: Contact;

  @ForeignKey(() => Whatsapp)
  @Column
  whatsappId: number;

  @BelongsTo(() => Whatsapp)
  whatsapp: Whatsapp;

  @ForeignKey(() => Queue)
  @Column
  queueId: number;

  @BelongsTo(() => Queue)
  queue: Queue;

  @ForeignKey(() => Step)
  @AllowNull(true)
  @Column
  stepId: number;

  @BelongsTo(() => Step)
  step: Step;

  @HasMany(() => Message)
  messages: Message[];

  @ForeignKey(() => Tenant)
  @Column(DataType.UUID)
  tenantId: number | string;

  @BelongsTo(() => Tenant)
  tenant: Tenant;

  @BelongsToMany(() => Tag, () => EntityTag, "entityId", "tagId")
  tags: Tag[];
}

export default Ticket;

File: bussines/internal/models/ticket.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Ticket struct {
	ID             int       `gorm:"primaryKey" json:"id"`
	Status         string    `gorm:"not null;default:'pending'" json:"status"`
	LastMessage    string    `gorm:"column:lastMessage" json:"lastMessage"`
	ContactID      int       `gorm:"column:contactId" json:"contactId"`
	UserID         *int      `gorm:"column:userId" json:"userId"`
	WhatsappID     int       `gorm:"column:whatsappId" json:"whatsappId"`
	IsGroup        bool      `gorm:"column:isGroup;not null;default:false" json:"isGroup"`
	UnreadMessages int       `gorm:"column:unreadMessages" json:"unreadMessages"`
	QueueID        *int      `gorm:"column:queueId" json:"queueId"`
	TenantID       uuid.UUID `gorm:"column:tenantId;type:uuid" json:"tenantId"`
	CreatedAt      time.Time `gorm:"column:createdAt" json:"createdAt"`
	UpdatedAt      time.Time `gorm:"column:updatedAt" json:"updatedAt"`

	// Relations
	Contact  Contact   `gorm:"foreignKey:ContactID" json:"contact,omitempty"`
	User     User      `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Messages []Message `gorm:"foreignKey:TicketID" json:"messages,omitempty"`
}

func (Ticket) TableName() string {
	return "Tickets"
}
